### PR TITLE
WIP: Adding support for LaunchTemplate config

### PR DESCRIPTION
**Pull request by preflightsiren**
Extended Public and Internal models, added License spec and Placement
information to aws.CreateLaunchTemplateInput calls.

### CODE REVIEW

File: controllers/provisioners/eks/scaling/launchtemplate.go [hunk 1]
```diff
@@ -237,6 +239,37 @@ func (lt *LaunchTemplate) Drifted(input *CreateConfigurationInput) bool {
 		drift = true
 	}
 
+	if len(latestVersion.LaunchTemplateData.LicenseSpecifications) != len(input.LicenseSpecifications) {
+		log.Info("detected drift", "reason", "Number of LicenseSpecifications has changed", "instancegroup", lt.OwnerName)
+		drift = true
+	}
+	for i, v := range latestVersion.LaunchTemplateData.LicenseSpecifications {
```
Comment (preflightsiren) [inline controllers/provisioners/eks/scaling/launchtemplate.go]: This is a bit of lazy drift check. The AWS API doesn't guarantee ordering of the license specs so we could end up in a situation where the API reorders the list causing the instance-manager to reconcile the node group.

File: controllers/provisioners/eks/helpers.go [hunk 2]
```diff
@@ -802,7 +803,7 @@ func (ctx *EksInstanceGroupContext) GetOverrides() []*autoscaling.LaunchTemplate
 	)
 	overrides := []*autoscaling.LaunchTemplateOverrides{}
 
-	if mixedPolicy.InstanceTypes != nil {
+	if mixedPolicy != nil && mixedPolicy.InstanceTypes != nil {
```
Comment (preflightsiren) [inline controllers/provisioners/eks/helpers.go:824]: nil check.

File: controllers/provisioners/eks/scaling/interface.go [hunk 3]
```diff
@@ -57,4 +57,12 @@ type CreateConfigurationInput struct {
 	Volumes               []v1alpha1.NodeVolume
 	UserData              string
 	SpotPrice             string
+	LicenseSpecifications []string
+	Placement             *LaunchTemplatePlacementInput
+}
+
+type LaunchTemplatePlacementInput struct {
```
Comment (preflightsiren) [inline controllers/provisioners/eks/scaling/interface.go:64]: Do we need to have a private model **and** a public model? I notice we just have `v1alpha1.NodeVolume` and no internal representation of a volume.

File: controllers/provisioners/eks/scaling/launchtemplate_test.go [hunk 4]
```diff
@@ -597,15 +599,101 @@ func TestLaunchTemplateDrifted(t *testing.T) {
...
-	for i, tc := range tests {
-		t.Logf("Test #%v", i)
-		ec2Mock.LaunchTemplates = []*ec2.LaunchTemplate{tc.launchTemplate}
-		lt, err := NewLaunchTemplate("", w, discoveryInput)
-		g.Expect(err).NotTo(gomega.HaveOccurred())
-		lt.LatestVersion = tc.latestVersion
-		result := lt.Drifted(tc.input)
-		g.Expect(result).To(gomega.Equal(tc.shouldDrift))
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
```
Comment (preflightsiren) [inline controllers/provisioners/eks/scaling/launchtemplate_test.go:690]: providing names makes it easier to find failing tests eg.
```
=== CONT  TestLaunchTemplateDrifted
    launchtemplate_test.go:653:
        Expected
            <bool>: true
        to equal
            <bool>: false
=== CONT  TestLaunchTemplateDrifted/NoDriftWhenLicenseSpecificationsSame
    testing.go:1052: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
--- FAIL: TestLaunchTemplateDrifted (0.00s)
    --- PASS: TestLaunchTemplateDrifted/NoDriftForEmptyInstanceProfile (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenLatestVersionUnset (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenImageIDChanges (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenInstanceTypeChanges (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenInstanceProfileArnChanges (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenSecurityGroupsChange (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenKeyNameChanges (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenUserDataChanges (0.00s)
    --- PASS: TestLaunchTemplateDrifted/ShouldDriftWhenVolumeSizeChanges (0.00s)
    --- FAIL: TestLaunchTemplateDrifted/NoDriftWhenLicenseSpecificationsSame (0.00s)
```

File: controllers/provisioners/eks/scaling/launchtemplate.go [hunk 5]
```diff
@@ -201,7 +203,11 @@ func (lt *LaunchTemplate) Drifted(input *CreateConfigurationInput) bool {
 		drift = true
 	}
 
-	if aws.StringValue(latestVersion.LaunchTemplateData.IamInstanceProfile.Arn) != input.IamInstanceProfileArn {
+	currentInstanceProfileArn := ""
+	if latestVersion.LaunchTemplateData.IamInstanceProfile != nil {
+		currentInstanceProfileArn = aws.StringValue(latestVersion.LaunchTemplateData.IamInstanceProfile.Arn)
+	}
+	if currentInstanceProfileArn != input.IamInstanceProfileArn {
```
Comment (preflightsiren) [inline controllers/provisioners/eks/scaling/launchtemplate.go:210]: nil check for `latestVersion.LaunchTemplateDate.IamInstanceProfile`

File: controllers/common/utils.go [hunk 6]
```diff
@@ -218,6 +218,12 @@ func StringPtr(str string) *string {
 }
 
 func StringSliceEquals(x, y []string) bool {
+	if x == nil {
+		x = []string{}
+	}
+	if y == nil {
+		y = []string{}
+	}
```
Comment (preflightsiren) [inline controllers/common/utils.go:226]: Adding nil checking for `x/y` removing the need to set IamInstanceProfile and SecurityGroupIds in tests.
It was failing as the arguments were `StringSliceEqual([]string{}, nil)` eg. https://github.com/eytan-avisror/instance-manager/pull/3/files#diff-6d52e99e4cfeaf5b9650f8e3992b44c55e4f2bff302917eb5fdbe2948c3104bcR218

Comment (eytan-avisror) [reply controllers/common/utils.go:226]: Good catch, i'll address it

**Pull request closed without merging**